Mathlib/CategoryTheory/Limits/Shapes/Fiberwise.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2025 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import Mathlib.CategoryTheory.FiberedCategory.Fiber
+import Mathlib.CategoryTheory.Limits.Shapes.Products
+
+/-!
+
+# Computing Colimits Fiberwise
+
+In this file, we consider category `J` equipped with a functor `F : J ⥤ D` to a discrete category
+`D`. Then the colimit of any diagram `diagram : J ⥤ C` can be computed fiberwise, using the
+following algorithm:
+
+1. For each `d : D`, construct a cocone over the restricted diagram `fiberInclusion F d ⋙ diagram`.
+2. Take a cofan of the values of these cocones over all `d : D`.
+
+## Main Results
+
+- `colimitOfFiber`: As above, given colimits on each restriction, and coproduct of the values, this
+  gives the colimit of the original diagram `diagram`.
+
+-/
+
+universe v v₁ v₂ u u₁ u₂
+
+open CategoryTheory Functor Category Fiber
+
+variable {J : Type u} {D : Type u₁}
+  [Category.{v} J] [Category.{v₁} D] [IsDiscrete D]
+  (F : J ⥤ D)
+  {C : Type u₂} [Category.{v₂} C] (diagram : J ⥤ C)
+
+namespace CategoryTheory
+
+namespace Functor
+
+theorem fiberCongr {j₁ j₂ : J} (f : j₁ ⟶ j₂) : F.obj j₁ = F.obj j₂ :=
+  IsDiscrete.eq_of_hom (F.map f)
+
+instance (d : D) {j₁ j₂ : F.Fiber d} (f : j₁.1 ⟶ j₂.1) : F.IsHomLift (𝟙 d) f :=
+  IsHomLift.of_fac _ _ _ j₁.2 j₂.2 (Subsingleton.elim _ _)
+
+/-- Casting a morphism in `J` to a morphism in the category `F.Fiber d`. -/
+def fiberPreimage {d : D} (j₁ j₂ : F.Fiber d) (f : j₁.1 ⟶ j₂.1) : j₁ ⟶ j₂ :=
+  homMk F d f
+
+/-- The inclusion functor from `F.Fiber d` to `J` is fully faithful when `D` is discrete. -/
+def fullyFaithfulFiber (d : D) : FullyFaithful (fiberInclusion (p := F) (S := d)) where
+  preimage {j₁ j₂} := F.fiberPreimage j₁ j₂
+
+instance (d : D) : Full (fiberInclusion (p := F) (S := d)) :=
+  (fullyFaithfulFiber F d).full
+
+@[elab_as_elim] lemma fiber_inductionOn {motive : ∀ {j₁ j₂ : J}, (j₁ ⟶ j₂) → Prop}
+    {j₁ j₂ : J} (f : j₁ ⟶ j₂)
+    (ih : ∀ d : D, ∀ {j₁ j₂ : F.Fiber d} (f : j₁ ⟶ j₂), motive f.1) :
+    motive f :=
+  ih _ (F.fiberPreimage (.mk (F.fiberCongr f)) (.mk rfl) f)
+
+end Functor
+
+namespace Limits
+
+open Functor
+
```

Comment:
Same comment about the names in this section: they are too generic given that they talk about a rather specific situation.

Reply:
done